app/scripts/modules/oracle/serverGroup/configure/serverGroupConfiguration.service.js

```diff
@@ -0,0 +1,112 @@
+  ACCOUNT_SERVICE,
+  NETWORK_READ_SERVICE,
+  SUBNET_READ_SERVICE,
+  SECURITY_GROUP_READER
+])
+  .factory('oraclebmcsServerGroupConfigurationService', function($q, networkReader, oraclebmcsSubnetReader, accountService, oraclebmcsImageReader, securityGroupReader) {
+
+    let oracle = 'oraclebmcs';
+
+    let getShapes = (image) => {
+      if (!image || !image.compatibleShapes) {
+        return [];
+      }
+      return _.map(image.compatibleShapes, function(shape) {
+        return { name: shape };
+      });
+    };
+
+    function configureCommand(application, command) {
+      let defaults = command || {};
+      let defaultCredentials = defaults.account || application.defaultCredentials.oraclebmcs || OracleBMCSProviderSettings.defaults.account;
+      let defaultRegion = defaults.region || application.defaultRegions.oraclebmcs || OracleBMCSProviderSettings.defaults.region;
+
+        return $q.all({
+            credentialsKeyedByAccount: accountService.getCredentialsKeyedByAccount(oracle),
+            networks: networkReader.listNetworksByProvider(oracle),
+            subnets: oraclebmcsSubnetReader.listSubnetsByProvider(oracle),
+            images: loadImages(),
+            availDomains: accountService.getAvailabilityZonesForAccountAndRegion(oracle, defaultCredentials, defaultRegion)
+        }).then(function(backingData) {
+            backingData.accounts = _.keys(backingData.credentialsKeyedByAccount);
+            backingData.filtered = {};
+            backingData.filtered.regions = [{name: 'us-phoenix-1'}];
+            backingData.filtered.availabilityDomains = _.map(backingData.availDomains, function(zone) {
+              return {name: zone};
+            });
+
+            backingData.filterSubnets = function() {
+              if(command.vpcId && command.availabilityDomain) {
+                return _.filter(backingData.subnets, {
+                  vcnId: command.vpcId,
+                  availabilityDomain: command.availabilityDomain
+                });
+              }
+              return backingData.subnets;
+            };
+
+            backingData.availabilityDomainOnChange = function() {
+              command.subnetId = null;
+              backingData.seclists = null;
+            };
+
+            backingData.vpcOnChange = function() {
+              command.subnetId = null;
+              backingData.seclists = null;
+            };
+
+            backingData.subnetOnChange = function() {
+              let subnet = _.find(backingData.subnets, {id : command.subnetId});
+
+              securityGroupReader.getAllSecurityGroups().then(function(allSecurityGroups) {
+                let mySecGroups = allSecurityGroups[command.account][oracle][command.region];
+                let secLists = [];
+                _.forEach(subnet.securityListIds, function(sid) {
+                  let sgRef = _.find(mySecGroups, {id: sid});
+                  securityGroupReader.getSecurityGroupDetails(
```

Comment:
If I'm reading this right, it could potentially get very expensive - it looks like make an ajax request for each security group in the subnet, and if there are more than a handful, it'll bog down the client.

If you don't absolutely need the details for each security group, I'd strongly suggest refactoring this.